File: algorithm/bfs/miro.py
#!/usr/bin/env python3
import sys
import queue

maps=[]
visited=[]
T_X=None
T_Y=None

if __name__ == '__main__':
    readl=sys.stdin.readline
    C,R=map(int,readl().strip().split())
    X, Y, T_X, T_Y=map(int,readl().strip().split())
    X-=1
    Y-=1
    T_X-=1
    T_Y-=1
    for i in range(R):
        maps.append([])
        visited.append([])
        line=readl().strip()
        maps[i]=[-1]*C
        visited[i]=[False]*C
        for j in range(C):
            maps[i][j]=line[j]
    q=queue.Queue()
    q.put_nowait((Y,X,0))
    dx=[0,0,-1,1] # x = left,right
    dy=[-1,1,0,0] # y = up, down
    while not q.empty():
        now_y, now_x, now_time=q.get_nowait()
        visited[now_y][now_x]=True
        if now_x == T_X and now_y == T_Y:
            print(now_time)
            break
        for i in range(0,4):
            nx=now_x+dx[i]
            ny=now_y+dy[i]
            if (nx < 0) or (nx >= C):
                continue
            if (ny < 0) or (ny >= R):
                continue
            if visited[ny][nx]==True: 
                continue
            if maps[ny][nx]=='1':
                continue
            q.put_nowait((ny,nx,now_time+1))
    exit(0)

File: algorithm/bfs/miro.c
#include <stdio.h>

int X,Y;
int sx,sy,ex,ey;
char map[102][102];
int visited[102][102];

struct QUEUE{
	int y,x,time;
};

struct QUEUE queue[10000];
int wp,rp;

void Enqueue(int y, int x,int time){
	queue[wp].y=y;
	queue[wp].x=x;
	queue[wp].time=time;
	wp++;
}

struct QUEUE Dequeue(){
	return queue[rp++];
}

int Empty(){
	return wp == rp;
}

int main(void){
	scanf("%d %d", &X, &Y);
	scanf("%d %d %d %d", &sx, &sy, &ex, &ey);
	for(int i=1; i<=Y; i++){
		scanf("%s", &map[i][1]);
	}
	int dx[]={0,0,1,-1};
	int dy[]={1,-1,0,0};

	Enqueue(sy, sx,0);
	visited[sy][sx]=1;
	while(!Empty()){
		struct QUEUE cur=Dequeue();
		if((cur.y==ey)&&(cur.x==ex)){
			printf("%d",cur.time);
		}
		for(int i=0; i<4;i++){
			int ny=cur.y+dy[i];
			int nx=cur.x+dx[i];
			if((ny<1)||(ny>Y)) continue;
			if((nx<1)||(nx>X)) continue;
			if(visited[ny][nx]==1) continue;
			if(map[ny][nx] != '0') continue;
			Enqueue(ny,nx,cur.time+1);
			visited[ny][nx]=1;
		}
	}
	return -1;
}
